### PR TITLE
Fix udp socket

### PR DESCRIPTION
**Pull request by ClawSeven**
Rebase new commits and replace #301.

### CODE REVIEW

File: src/libos/src/net/socket_file.rs [hunk 1]
```diff
@@ -205,6 +214,10 @@ impl SocketFile {
                         return_errno!(EINVAL, "not support IPv6, yet");
                     }
                 };
+                if protocol != SocketProtocol::IPPROTO_IP && protocol != SocketProtocol::IPPROTO_UDP
```
Comment (volcano0dr) [inline src/libos/src/net/socket_file.rs:217]:
```
if domain != Domain::Netlink && protocol != SocketProtocol::IPPROTO_IP && protocol != SocketProtocol::IPPROTO_UDP {
    return_errno!(EINVAL, "unsupported protocol");
}
```

File: src/libos/crates/async-io/src/socket/addr/mod.rs [hunk 2]
```diff
@@ -31,6 +31,8 @@ pub trait Addr: Clone + Debug + Default + PartialEq + Send + Sync {
     fn to_c_storage(&self) -> (libc::sockaddr_storage, usize);
 
     fn as_any(&self) -> &dyn Any;
+
+    fn is_default(&self) -> bool;
```
Comment (volcano0dr) [inline src/libos/crates/async-io/src/socket/addr/mod.rs:35]: For IP address, `is_any_addr` is meaningful. For UDS and Netlink, I think it is not necessary to provide the `is_default` method.

File: src/libos/crates/host-socket/src/datagram/netlink.rs [hunk 3]
```diff
@@ -117,10 +121,11 @@ impl<A: Addr, R: Runtime> NetlinkSocket<A, R> {
         flags: SendFlags,
     ) -> Result<usize> {
         let res = if addr.is_some() {
-            self.sender.sendmsg(bufs, addr, flags).await
+            self.sender.sendmsg(bufs, addr.unwrap(), flags, None).await
         } else {
             let peer = self.common.peer_addr();
-            self.sender.sendmsg(bufs, peer.as_ref(), flags).await
+            // Need to check peer addr.
```
Comment (volcano0dr) [inline src/libos/crates/host-socket/src/datagram/netlink.rs:127]: Need to check peer addr

File: src/libos/crates/host-socket/src/datagram/generic.rs [hunk 4]
```diff
@@ -172,36 +210,49 @@ impl<A: Addr, R: Runtime> DatagramSocket<A, R> {
...
         let state = self.state.read().unwrap();
         if addr.is_none() && !state.is_connected() {
-            return_errno!(ENOTCONN, "the socket is not connected");
+            return_errno!(EDESTADDRREQ, "Destination address required");
         }
 
         let res = if addr.is_some() {
             drop(state);
-            self.sender.sendmsg(bufs, addr, flags).await
+            self.sender
```
Comment (volcano0dr) [inline src/libos/crates/host-socket/src/datagram/generic.rs:244]:
```
let res = if let Some(addr) = addr {
    drop(state);
    self.sender
        .sendmsg(bufs, addr, flags, control)
        .await
} else {
    let peer = self.common.peer_addr();
    drop(state);
    if let Some(peer) = peer.as_ref() {
        self.sender
            .sendmsg(bufs, peer, flags, control)
            .await
    } else {
        return_errno!(EDESTADDRREQ, "Destination address required");
    }
};
```

File: src/libos/crates/host-socket/src/datagram/sender.rs [hunk 5]
```diff
@@ -68,26 +104,41 @@ impl<A: Addr, R: Runtime> Sender<A, R> {
...
 
-    let (c_addr_storage, c_addr_len) = if let Some(addr) = addr {
-        addr.to_c_storage()
-    } else {
-        let storage: libc::sockaddr_storage = unsafe { std::mem::zeroed() };
-        (storage, 0)
-    };
+    let (c_addr_storage, c_addr_len) = addr.to_c_storage();
+
+    // let (c_addr_storage, c_addr_len) = if let Some(addr) = addr {
```
Comment (volcano0dr) [inline src/libos/crates/host-socket/src/datagram/sender.rs:122]: Please remove the deprecated code.

File: src/libos/crates/host-socket/src/datagram/receiver.rs [hunk 6]
```diff
@@ -314,6 +410,7 @@ struct RecvReq {
     msg: libc::msghdr,
     iovec: libc::iovec,
     addr: libc::sockaddr_storage,
+    control: *mut c_void,
```
Comment (volcano0dr) [inline src/libos/crates/host-socket/src/datagram/receiver.rs:413]: I think the control field is useless.

File: src/libos/crates/host-socket/src/datagram/receiver.rs [hunk 7]
```diff
@@ -68,20 +93,57 @@ impl<A: Addr, R: Runtime> Receiver<A, R> {
...
+                }
+                let control_buf_len = if control.is_some() {
+                    control.as_ref().unwrap().len()
+                } else {
+                    0
+                };
+                assert!(msg_controllen >= control_buf_len);
+
+                if msg_controllen > 0 {
+                    control
```
Comment (volcano0dr) [inline src/libos/crates/host-socket/src/datagram/receiver.rs:121]: The `MSG_CTRUNC` flag should be set when some control data is discarded.

File: src/libos/crates/async-io/src/socket/addr/ipv4.rs [hunk 8]
```diff
@@ -20,6 +20,12 @@ impl Addr for Ipv4SocketAddr {
         if c_addr_len > std::mem::size_of::<libc::sockaddr_storage>() {
             return_errno!(EINVAL, "address length is too large");
         }
+        // TODO: Use addr length more specifically.
+        // The hard code value of 16 is the length of IN_ADDR_ANY.
+        // https://en.wikipedia.org/wiki/IPv4
+        if c_addr_len < 16 {
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply src/libos/crates/async-io/src/socket/addr/ipv4.rs]: Done

File: src/libos/crates/async-io/src/socket/addr/ipv4.rs [hunk 9]
```diff
@@ -74,6 +85,11 @@ impl Ipv4SocketAddr {
     pub fn set_port(&mut self, new_port: u16) {
         self.port = new_port;
     }
+
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply src/libos/crates/async-io/src/socket/addr/ipv4.rs]: Done

File: src/libos/crates/async-io/src/socket/addr/ipv6.rs [hunk 10]
```diff
@@ -25,6 +25,12 @@ impl Addr for Ipv6SocketAddr {
         if c_addr_len > std::mem::size_of::<libc::sockaddr_storage>() {
             return_errno!(EINVAL, "address length is too large");
         }
+        // TODO: Use addr length more specifically.
+        // The hard code value of 16 is the length of IN_ADDR_ANY.
+        // https://en.wikipedia.org/wiki/IPv6
+        if c_addr_len < 16 {
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply src/libos/crates/async-io/src/socket/addr/ipv6.rs]: Done

File: src/libos/crates/async-io/src/socket/addr/ipv6.rs [hunk 11]
```diff
@@ -90,6 +101,11 @@ impl Ipv6SocketAddr {
     pub fn set_port(&mut self, new_port: u16) {
         self.port = new_port;
     }
+
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply src/libos/crates/async-io/src/socket/addr/ipv6.rs]: Done

File: src/libos/src/util/mem_util.rs [hunk 12]
```diff
@@ -27,6 +27,12 @@ pub mod from_user {
 
     /// Check the readonly array is within the readable memory of the user process
     pub fn check_array<T>(user_buf: *const T, count: usize) -> Result<()> {
+        // If user_buf pointer is null and count is zero, it is ok.
+        if user_buf.is_null() && count == 0 {
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply src/libos/src/util/mem_util.rs]: Done

File: src/libos/src/net/socket_file.rs (hunk 1 above)
Comment (ClawSeven) [reply src/libos/src/net/socket_file.rs]: Done

File: src/libos/crates/host-socket/src/datagram/sender.rs (hunk 5 above)
Comment (ClawSeven) [reply src/libos/crates/host-socket/src/datagram/sender.rs]: Done

File: src/libos/crates/async-io/src/socket/addr/mod.rs (hunk 2 above)
Comment (ClawSeven) [reply src/libos/crates/async-io/src/socket/addr/mod.rs:35]: Done, I changed the UDS and Netlink `is_default()` method to `unimplemented!`

File: src/libos/crates/host-socket/src/datagram/receiver.rs [hunk 13]
```diff
@@ -46,7 +53,22 @@ impl<A: Addr, R: Runtime> Receiver<A, R> {
             }
             let events = self.common.pollee().poll(mask, None);
             if events.is_empty() {
-                poller.as_ref().unwrap().wait().await?;
+                let ret = poller
+                    .as_ref()
+                    .unwrap()
+                    .wait_timeout(self.common.recv_timeout().as_mut())
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply src/libos/crates/host-socket/src/datagram/receiver.rs]: Done

File: src/libos/crates/host-socket/src/datagram/netlink.rs (hunk 3 above)
Comment (ClawSeven) [reply src/libos/crates/host-socket/src/datagram/netlink.rs]: Done

File: src/libos/crates/host-socket/src/datagram/generic.rs (hunk 4 above)
Comment (ClawSeven) [reply src/libos/crates/host-socket/src/datagram/generic.rs]: Done

File: src/libos/crates/host-socket/src/datagram/receiver.rs (hunk 6 above)
Comment (ClawSeven) [reply src/libos/crates/host-socket/src/datagram/receiver.rs]: Done

File: src/libos/crates/host-socket/src/datagram/receiver.rs (hunk 7 above)
Comment (ClawSeven) [reply src/libos/crates/host-socket/src/datagram/receiver.rs:126]: Done

File: src/libos/src/net/syscalls.rs [hunk 14]
```diff
@@ -1,5 +1,6 @@
 use std::convert::TryFrom;
 use std::mem::MaybeUninit;
+use std::ptr::{self};
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply src/libos/src/net/syscalls.rs]: Done

File: src/libos/src/net/syscalls.rs [hunk 15]
```diff
@@ -306,9 +308,18 @@ pub async fn do_recvmsg(fd: c_int, msg_mut_ptr: *mut libc::msghdr, flags: c_int)
         }
     }
 
-    // Clear msghdr flags to 0 after recvmsg
-    // FIXME: Several kinds of flags are set by the kernel, e.g. MSG_TRUNC. Support these flags in the future.
-    msg.msg_flags = 0;
+    // update msghdr
```
[the remark this answers is not in the document]
Comment (ClawSeven) [reply src/libos/src/net/syscalls.rs]: Done

**Review by volcano0dr — APPROVED** [review]
LGTM

**Pull request merged**